src/invariant/OrderBookIsNotCrossed.cpp

```diff
@@ -0,0 +1,225 @@
+
+#include "invariant/OrderBookIsNotCrossed.h"
+#include "invariant/InvariantManager.h"
+#include "ledger/InternalLedgerEntry.h"
+#include "ledger/LedgerTxn.h"
+#include "main/Application.h"
+#include "util/Logging.h"
+#include "util/XDROperators.h"
+#include "xdr/Stellar-ledger-entries.h"
+#include <fmt/format.h>
+#include <functional>
+#include <xdrpp/printer.h>
+
+namespace stellar
+{
+
+namespace
+{
+static AssetPairSet
+extractAssetPairs(LedgerTxnDelta const& ltxd)
+{
+    AssetPairSet assets;
+    for (auto const& entry : ltxd.entry)
+    {
+        if (entry.first.type() == InternalLedgerEntryType::LEDGER_ENTRY &&
+            entry.first.ledgerKey().type() == OFFER && entry.second.current)
+        {
+            auto const& offer =
+                entry.second.current->ledgerEntry().data.offer();
+            auto const assetPair =
+                offer.selling < offer.buying
+                    ? std::make_pair(offer.selling, offer.buying)
+                    : std::make_pair(offer.buying, offer.selling);
+
+            assets.insert(assetPair);
+        }
+    }
+
+    return assets;
+}
+
+static double
+price(OfferEntry const& offer)
+{
+    return double(offer.price.n) / double(offer.price.d);
+}
+
+static std::string
+checkCrossed(Asset const& a, Asset const& b, OrderBook const& orderBook)
+{
+    // if either side of order book for asset pair empty or does not yet exist,
+    // order book cannot be crossed
+    auto const iterAB = orderBook.find({a, b});
+    auto const iterBA = orderBook.find({b, a});
+    if (iterAB == orderBook.end() || iterBA == orderBook.end())
+    {
+        return {};
+    }
+
+    auto const& asks = iterAB->second;
+    auto const& bids = iterBA->second;
+    if (asks.empty() || bids.empty())
+    {
+        return {};
+    }
+
+    // check if crossed
+    auto lowestAsk = price(*asks.cbegin());
+    auto highestBid = 1 / price(*bids.cbegin());
+
+    // Allow highestBid to equal (not exceed) lowestAsk if at least one of the
+    // offers is passive.
+    if (lowestAsk == highestBid)
+    {
+        auto existsEqualPriceNonPassive =
+            [](auto const targetPrice, Orders const& offers, auto& found) {
+                for (auto const& offer : offers)
+                {
+                    if (price(offer) != targetPrice)
+                    {
+                        return false;
+                    }
+                    if ((offer.flags & PASSIVE_FLAG) == 0)
+                    {
+                        found = offer;
+                        return true;
+                    }
+                }
+                return false;
+            };
+
+        OfferEntry ask, bid;
+        auto nonPassiveAskAtSamePrice =
+            existsEqualPriceNonPassive(lowestAsk, asks, ask);
+        auto nonPassiveBidAtSamePrice =
+            existsEqualPriceNonPassive(1 / lowestAsk, bids, bid);
+
+        if (nonPassiveAskAtSamePrice && nonPassiveBidAtSamePrice)
```

Comment:
Based on all of my comments here, I believe that there is actually an O(1) way to perform this check. Your current implementation is O(n) so this would be a big improvement.

Reply:
My comment above (https://github.com/stellar/stellar-core/pull/2959#discussion_r595418524) explains why I currently do not see how we can avoid searching a list of offers at a price that exactly equally crosses some pre-existing offer to find out whether any of them is not passive, if we are to catch a bug where there is such an offer.  (Unless we were to check and update state with each individual `Operation` in `checkOnOperationApply`, rather than extracting all the asset pairs and batching the updates and then doing a single check, but I would expect that to be slowing down more common cases to speed up this rarer one.)  I might be misreading the semantics of passive offers, though!

Reply:
Despite your clarifications about how this implementation works, I still believe this can be done with O(1) complexity.

Reply:
I eventually caught on to Jon's clever idea here:  extend `OfferEntryCmp` to sort non-passive orders before passive orders at the same price.  Then we know that if there's _any_ non-passive order in the list of offers at the same price, then one of the non-passive orders will be at the front.  That allows us to check only the fronts of each list.